ml/Database.cc

```diff
@@ -38,6 +38,39 @@ const char *ml::Database::SQL_SELECT_ANOMALY_EVENTS =
     "   after >= ?4 AND"
     "   before <= ?5;";
 
+const char *ml::Database::SQL_CREATE_ANOMALY_RATE_INFO_TABLE =
+    "CREATE TABLE IF NOT EXISTS anomaly_rate_info( "
+    "     host_id text NOT NULL, "
+    "     after int NOT NULL, "
+    "     before int NOT NULL, "
+    "     anomaly_rates text "
+    ");";
+
+const char *ml::Database::SQL_INSERT_BULK_ANOMALY_RATE_INFO =
+    "INSERT INTO anomaly_rate_info( "
+    "     host_id, after, before, anomaly_rates) "
+    " VALUES (?1, ?2, ?3, ?4);";
+   
+const char *ml::Database::SQL_SELECT_ANOMALY_RATE_INFO =
+    "SELECT dimension_id, AVG(anomaly_percentage) "
+    "  FROM (SELECT json_extract(j.value, '$[1]') AS dimension_id, "
+    "  json_extract(j.value, '$[0]') AS anomaly_percentage "
+    "  FROM anomaly_rate_info AS ari, json_each(ari.anomaly_rates) AS j " 
+    "  WHERE ari.host_id == ?1 AND ari.after >= ?2 AND ari.before <= ?3 "
+    "  AND json_valid(ari.anomaly_rates)) "
+    "  GROUP BY dimension_id;";
+
+const char *ml::Database::SQL_SHRINK_ANOMALY_RATE_INFO =
+    "DELETE FROM anomaly_rate_info "
+    " WHERE after NOT IN ( "
+    " SELECT after FROM ( "
+    " SELECT after FROM anomaly_rate_info "
+    " ORDER BY after DESC LIMIT ?1 ) foo);";
+
+const char *ml::Database::SQL_REMOVE_OLD_ANOMALY_RATE_INFO =
+    "DELETE FROM anomaly_rate_info "
+    " WHERE before <= ?1;";
```

Comment:
Not sure why we need two statements in this case. I think just the `SQL_REMOVE_OLD_ANOMALY_RATE_INFO` statement will suffice, right?

Reply:
I am leaving two strategies, and I think somehow we will benefit from both, none contradicts or troubles the other. One is in terms of controlling the number of rows, so, the volume of the database, or the file size, and the other is the age of the data, say even if the file size has not passed beyond the threshold, we can still save space by getting rid of very old data.